exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/dml/UpsertTests.kt

```diff
@@ -325,6 +333,59 @@ class UpsertTests : DatabaseTestsBase() {
         }
     }
 
+    @Test
+    fun testUpsertWithManualUpdateUsingInsertValues() {
+        val tester = object : Table("tester") {
+            val id = integer("id").uniqueIndex()
+            val word = varchar("name", 64)
+            val count = integer("count").default(1)
+        }
+
+        withTables(excludeSettings = TestDB.ALL_H2_V1, tester) {
+            tester.insert {
+                it[id] = 1
+                it[word] = "Word A"
+            }
+            assertEquals(1, tester.selectAll().single()[tester.count])
+
+            // H2_Mysql & H2_Mariadb syntax does not allow VALUES() syntax to come first in complex expression
+            // Syntax must be column=(1 + VALUES(column)), not column=(VALUES(column) + 1)
+            tester.upsert(
+                onUpdate = { listOf(tester.count to intLiteral(100).times(tester.count.asForInsert())) }
+            ) {
+                it[id] = 1
+                it[word] = "Word B"
+                it[count] = 9
+            }
+            val result = tester.selectAll().single()
+            assertEquals(900, result[tester.count])
+
+            val newWords = listOf(
+                Triple(2, "Word B", 2),
+                Triple(1, "Word A", 3),
+                Triple(3, "Word C", 4)
+            )
+            tester.batchUpsert(
+                newWords,
+                onUpdate = {
+                    listOf(
```

Comment:
If it's breaking change anyway, would it better to make here a builder like other insert/update/upsert builders, it looks like:

```kotlin
onUpdate = {
  it[tester.word] = concat(tester.word, stringLiteral(" || "), tester.count.asForInsert())
  it[tester.count] = intLiteral(1).plus(tester.count.asForInsert())
}
```

For me it's always painfull to check if I should write in the builder `this[column] = value` or `it[column] = value`, with this change we will have also `listOf(column to value)` (don't know if we use such patter somewhere else)

Probably we could try to use for setters only `it`, and from `this` take only contextual methods (like `asForInsert()` in this context).

Reply:
Probably, to avoid breaking change, we could deprecate this method, and create another one where `onUpdate` is already a lambda

Reply:
I can try to make a new builder as I already tried using the existing builder constructs and had issue with capturing the set arguments within the correct statement.

About the deprecation, could you let me know if/how it is possible? When I tried to deprecate the old and add the new with a lambda, because of the default arguments, it causes all usages to fail with overload resolution ambiguity.

Reply:
I got the point about deprecations...

Not sure that it's legal, but removing default parameter from `onUpdate` of old function would solve that problem:

```kotlin
fun <T : Table, E : Any> T.batchUpsert(
    data: Iterable<E>,
    vararg keys: Column<*>,
    onUpdate: List<Pair<Column<*>, Expression<*>>>?,
    onUpdateExclude: List<Column<*>>? = null,
    where: (SqlExpressionBuilder.() -> Op<Boolean>)? = null,
    shouldReturnGeneratedValues: Boolean = true,
    body: BatchUpsertStatement.(E) -> Unit
): List<ResultRow> {
    return batchUpsert(data.iterator(), onUpdate?.let { { it } }, onUpdateExclude, where, shouldReturnGeneratedValues, keys = keys, body = body)
}
```

I'm trying to understand would it break somebody, and imagination says that should not.. but I can not be 100% sure. All the calls with `onUpdate` should continue to call the old deprecation function. All the calls without `onUpdate` should start to use new one automatically... is that logic?)

Reply:
In merge command I was creating `InsertStatement` and `UpdateStatement` as receivers, and took the arguments from there, not sure if it's applicable here.. I still have an idea to reuse merge command for upsert if it's possible, but last time I face the problem, that I needed `currentDialect` in the moment it's not available yet, but I hope I'll make one more attempt later